### ISSUE TITLE
Loky-backed parallel loops cannot be nested below threads, setting n_jobs=1

### ISSUE THREAD

**Issue by jl-hk** [joblib/joblib#978]
I am getting 'Loky-backed parallel loops cannot be nested below threads, setting n_jobs=1
  **self._backend_args)' when loky backend is placed inside a threading backend task. May I know if it is possible to do so?

I am running joblib 0.13.2 in python3.7 with macOS.

Not sure if this is the same as #180 . Since the error message is slightly different from #180's, I am creating a new issue here.

**Comment by ogrisel** [comment]
Can you please provide a minimal reproducing code snippet?

We don't want multi-processing parallel calls nested under thread-based parallel calls. Why would you want this in the first place?

**Comment by jl-hk** [comment]
I was originally trying to find a work around solution for `SIGSEGV11` error which happens quite often when I do 2 level nested loky calls with numpy function inside. Later I found that adding 
`with parallel_backend("loky", inner_max_num_threads=1):`
solves the problem and I can use 2 level nested loky calls again.

Thank you for pointing out that we should not do multi-processing calls nested under thread calls. I will close this issue if no one has the same problem.

**Comment by GaelVaroquaux** [comment]
If you see a way to improve our documentation, we would love a pull
request there.

Thanks!

**Comment by chrisspen** [comment]
>We don't want multi-processing parallel calls nested under thread-based parallel calls. Why would you want this in the first place?

Because threading in Python is horrible inefficient due to the GIL, whereas multiprocessing in Python is very efficient on systems that have multiple cores.

**Comment by troehling** [comment]
I am getting a very similar warning when running with multiiprocessing: `Loky-backed parallel loops cannot be called in a multiprocessing, setting n_jobs=1`. I use scikit-learn's DBSCAN algorithm with n_jobs > 1. Is this message intended? I assume that the computation is running only on a single core in this case. Is there any possibility how I could get this to work or do I have to either launch the algorithm in the main process of implement multiprocessing myself?

**Comment by tomMoral** [comment]
Yeah this message is intended. If you wish to have working `loky` backend under `multiprocessing`, I guess you would need to use `multiprocessing.set_start_method('loky')`, to have `LokyProcess` and not `ForkProcess` in your upper level multiprocessing.

Note that this issue is quite old and closed so feel free to open a new one to keep track of your problem.

**Comment by troehling** [comment]
I have tried this but it does not seem to be working. Thank you very much!

**Comment by linar-jether** [comment]
If anyone else encounters this issue, i'm copying my comment from https://github.com/celery/celery/issues/4551#issuecomment-886653650.

This is a simple workaround that monkey-patches joblib to allow for nested concurrency (inside Celery or any other pool).
**Use with care** as in some case this can create many processes and cause oversubscription.

```python
def _patch_joblib_loky_backend():
    import joblib._parallel_backends
    from joblib._parallel_backends import mp, cpu_count
    
    def effective_n_jobs(self, n_jobs):
        """Determine the number of jobs which are going to run in parallel"""
        if n_jobs == 0:
            raise ValueError('n_jobs == 0 in Parallel has no meaning')
        elif mp is None or n_jobs is None:
            # multiprocessing is not available or disabled, fallback
            # to sequential mode
            return 1
        elif n_jobs < 0:
            n_jobs = max(cpu_count() + 1 + n_jobs, 1)
        return n_jobs
    
    # Monkey-patch to allow daemonic thread to spawn processes
    joblib._parallel_backends.LokyBackend.effective_n_jobs = effective_n_jobs

# This works as a celery task, properly using nested concurrency
def multiprocessing_joblib_test_task(items):
    _patch_joblib_loky_backend()
    from joblib import Parallel, delayed, parallel_backend
    with parallel_backend("loky"):
        with Parallel(n_jobs=2) as pool:
            return pool(delayed(lambda x: x + 1)(x) for x in items)
```

**Comment by R-N** [comment]
> Yeah this message is intended. If you wish to have working `loky` backend under `multiprocessing`, I guess you would need to use `multiprocessing.set_start_method('loky')`, to have `LokyProcess` and not `ForkProcess` in your upper level multiprocessing.
> 
> Note that this issue is quite old and closed so feel free to open a new one to keep track of your problem.

Hello, I tried this and it says "ValueError: cannot find context for 'loky'". Am I missing something?

**Comment by ogrisel** [comment]
Please open a dedicated issue with a minimal reproducing example.
